### PR TITLE
feat(network): add datasource `cloudavenue_network_routed`

### PR DESCRIPTION
**Pull request by valentin-dubreuil**
### Description of your changes

Add datasource `cloudavenue_network_routed`

If you submit change in the provider code, please make sure to:

- [x] Write or modify examples in `examples/` directory
- [x] Write or modify acceptance tests
- [x] Run `make generate` to ensure the doc was updated properly

### How has this code been tested
```
=== RUN   TestAccNetworkRoutedDataSource
--- PASS: TestAccNetworkRoutedDataSource (4.90s)
PASS
ok      github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/tests/network     4.912s
```

### CODE REVIEW

File: internal/provider/network/routed_datasource.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	client *client.CloudAvenue
+}
+
+type networkRoutedDataSourceModel struct {
+	ID            types.String `tfsdk:"id"`
+	Name          types.String `tfsdk:"name"`
+	Description   types.String `tfsdk:"description"`
+	EdgeGatewayID types.String `tfsdk:"edge_gateway_id"`
+	InterfaceType types.String `tfsdk:"interface_type"`
+	Gateway       types.String `tfsdk:"gateway"`
```
Comment (azrod) [inline internal/provider/network/routed_datasource.go:47]: Add `EdgeGatewayName` in addition to `EdgeGatewayID`

File: internal/provider/network/routed_datasource.go
Comment (azrod) [inline internal/provider/network/routed_datasource.go:1]: Warning use latest version of `terraform-templates`

File: internal/provider/network/routed_datasource.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	PrefixLength  types.Int64  `tfsdk:"prefix_length"`
+	DNS1          types.String `tfsdk:"dns1"`
+	DNS2          types.String `tfsdk:"dns2"`
+	DNSSuffix     types.String `tfsdk:"dns_suffix"`
+	StaticIPPool  types.Set    `tfsdk:"static_ip_pool"`
+}
+
+func (d *networkRoutedDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_" + "network_routed"
```
Comment (azrod) [inline internal/provider/network/routed_datasource.go:56]:
```suggestion
	resp.TypeName = req.ProviderTypeName + "_" + categoryName + "_routed"
```

File: internal/provider/network/routed_datasource.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+func (d *networkRoutedDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
```
Comment (azrod) [inline internal/provider/network/routed_datasource.go:59]: Refactor your schema and using SuperSchema

File: internal/provider/network/routed_datasource.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	_ datasource.DataSourceWithConfigure = &networkRoutedDataSource{}
+)
+
+func NewNetworkRoutedDataSource() datasource.DataSource {
+	return &networkRoutedDataSource{}
+}
+
+type networkRoutedDataSource struct {
+	client *client.CloudAvenue
+}
```
Comment (azrod) [inline internal/provider/network/routed_datasource.go:39]: Add your necessary Meta Object (Org, Network, etc)

**Review by azrod — CHANGES_REQUESTED** [review]

File: internal/provider/common/network/schema.go [hunk 5]
```diff
@@ -245,15 +245,15 @@ func GetSchema(opts ...networkSchemaOpts) superschema.Schema {
 	switch params.typeNetwork {
 	case NAT_ROUTED:
 		// Add routed network specific attributes to the schema
-		_schema.Common.MarkdownDescription = "Provides a Cloud Avenue VDC routed Network."
-		_schema.Resource.MarkdownDescription = "This can be used to create, modify, and delete VDC routed networks."
+		_schema.Resource.MarkdownDescription = "Provides a Cloud Avenue VDC routed Network. This can be used to create, modify, and delete VDC routed networks."
```
Comment (eremid) [inline internal/provider/common/network/schema.go:248]: VDC => vDC

File: internal/provider/network/routed_datasource.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+			resp.Diagnostics.Append(diag)
+			return
+		}
+		orgNetwork, err = d.org.GetOpenApiOrgVdcNetworkByNameAndOwnerId(data.Name.ValueString(), *parentID)
+	}
+
+	if err != nil {
+		if govcd.ContainsNotFound(err) {
+			tflog.Debug(ctx, "Network not found, removing resource from state")
+			resp.State.RemoveResource(ctx)
```
Comment (eremid) [inline internal/provider/network/routed_datasource.go:121]: `return`

**Review by eremid — APPROVED** [review]

**Review by azrod — APPROVED** [review]

File: internal/provider/network/common_network.go [hunk 7]
```diff
@@ -46,3 +64,36 @@ func GetParentEdgeGatewayID(org org.Org, edgeGatewayID string) (*string, diag.Di
...
+		EdgeGatewayName: types.StringValue(network.OpenApiOrgVdcNetwork.Connection.RouterRef.Name),
+		InterfaceType:   types.StringValue(network.OpenApiOrgVdcNetwork.Connection.ConnectionType),
+		Gateway:         types.StringValue(network.OpenApiOrgVdcNetwork.Subnets.Values[0].Gateway),
+		PrefixLength:    types.Int64Value(int64(network.OpenApiOrgVdcNetwork.Subnets.Values[0].PrefixLength)),
+		DNS1:            utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSServer1),
+		DNS2:            utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSServer2),
+		DNSSuffix:       utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSSuffix),
+	}
+	return nrm
+}
```
Comment (azrod) [inline internal/provider/network/common_network.go:99]:
```suggestion
func SetDataToNetworkRoutedModel(network *govcd.OpenApiOrgVdcNetwork) networkRoutedModel {
	return networkRoutedModel{
		ID:              types.StringValue(network.OpenApiOrgVdcNetwork.ID),
		Name:            types.StringValue(network.OpenApiOrgVdcNetwork.Name),
		Description:     utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Description),
		EdgeGatewayID:   types.StringValue(network.OpenApiOrgVdcNetwork.Connection.RouterRef.ID),
		EdgeGatewayName: types.StringValue(network.OpenApiOrgVdcNetwork.Connection.RouterRef.Name),
		InterfaceType:   types.StringValue(network.OpenApiOrgVdcNetwork.Connection.ConnectionType),
		Gateway:         types.StringValue(network.OpenApiOrgVdcNetwork.Subnets.Values[0].Gateway),
		PrefixLength:    types.Int64Value(int64(network.OpenApiOrgVdcNetwork.Subnets.Values[0].PrefixLength)),
		DNS1:            utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSServer1),
		DNS2:            utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSServer2),
		DNSSuffix:       utils.StringValueOrNull(network.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSSuffix),
	}
}
```

File: internal/provider/network/common_network.go [hunk 8]
```diff
@@ -46,3 +64,36 @@ func GetParentEdgeGatewayID(org org.Org, edgeGatewayID string) (*string, diag.Di
...
+		for _, ipRange := range network.OpenApiOrgVdcNetwork.Subnets.Values[0].IPRanges.Values {
+			ipPool := staticIPPool{
+				StartAddress: types.StringValue(ipRange.StartAddress),
+				EndAddress:   types.StringValue(ipRange.EndAddress),
+			}
+			ipPools = append(ipPools, ipPool)
+		}
+	}
+	return ipPools
+}
```
Comment (azrod) [inline internal/provider/network/common_network.go:81]:
```suggestion
func GetIPRanges(network *govcd.OpenApiOrgVdcNetwork) []staticIPPool {
	ipPools := []staticIPPool{}

		for _, ipRange := range network.OpenApiOrgVdcNetwork.Subnets.Values[0].IPRanges.Values {
			ipPool := staticIPPool{
				StartAddress: types.StringValue(ipRange.StartAddress),
				EndAddress:   types.StringValue(ipRange.EndAddress),
			}
			ipPools = append(ipPools, ipPool)
		}
	return ipPools
}
```

File: internal/provider/network/routed_datasource.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+//	EdgeGatewayID   types.String `tfsdk:"edge_gateway_id"`
+//	EdgeGatewayName types.String `tfsdk:"edge_gateway_name"`
+//	InterfaceType   types.String `tfsdk:"interface_type"`
+//	Gateway         types.String `tfsdk:"gateway"`
+//	PrefixLength    types.Int64  `tfsdk:"prefix_length"`
+//	DNS1            types.String `tfsdk:"dns1"`
+//	DNS2            types.String `tfsdk:"dns2"`
+//	DNSSuffix       types.String `tfsdk:"dns_suffix"`
+//	StaticIPPool    types.Set    `tfsdk:"static_ip_pool"`
+//}
```
Comment (azrod) [inline internal/provider/network/routed_datasource.go:49]:
```suggestion
```

File: internal/provider/network/routed_datasource.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	// 	Description:     types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Description),
+	// 	EdgeGatewayID:   types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Connection.RouterRef.ID),
+	// 	EdgeGatewayName: types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Connection.RouterRef.Name),
+	// 	InterfaceType:   types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Connection.ConnectionType),
+	// 	Gateway:         types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Subnets.Values[0].Gateway),
+	// 	PrefixLength:    types.Int64Value(int64(orgNetwork.OpenApiOrgVdcNetwork.Subnets.Values[0].PrefixLength)),
+	// 	DNS1:            types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSServer1),
+	// 	DNS2:            types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSServer2),
+	// 	DNSSuffix:       types.StringValue(orgNetwork.OpenApiOrgVdcNetwork.Subnets.Values[0].DNSSuffix),
+	// }
```
Comment (azrod) [inline internal/provider/network/routed_datasource.go:140]:
```suggestion
```

File: internal/tests/network/routed_datasource_test.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	tests "github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/tests/common"
+)
+
+//go:generate go run github.com/FrangipaneTeam/tf-doc-extractor@latest -filename $GOFILE -example-dir ../../../examples -test
+const testAccNetworkRoutedDataSourceConfig = `
+resource "cloudavenue_network_routed" "example" {
+	name = "ExampleNetworkRouted"
+	gateway       = "192.168.10.254"
+	prefix_length = 24
+	edge_gateway_id = "urn:vcloud:gateway:dde5d31a-2f32-43ef-b3b3-127245958298"
```
Comment (azrod) [inline internal/tests/network/routed_datasource_test.go:18]: Use datasource

**Review by azrod — CHANGES_REQUESTED** [review]

**Review by azrod — APPROVED** [review]

**Pull request merged**